views/api.py

```diff
@@ -30,11 +32,69 @@
 
 if TYPE_CHECKING:
     from core import Application
+    from types_.config import Github
+    from types_.github import PostGist
+
+
+DISCORD_TOKEN_REGEX: re.Pattern[str] = re.compile(r"[a-zA-Z0-9_-]{23,28}\.[a-zA-Z0-9_-]{6,7}\.[a-zA-Z0-9_-]{27,}")
 
 
 class APIView(starlette_plus.View, prefix="api"):
-    def __init__(self, app: Application) -> None:
+    def __init__(self, app: Application, *, github_config: Github | None) -> None:
         self.app: Application = app
+        self._handling_tokens = bool(self.app.session and github_config)
+
+        if self._handling_tokens:
+            assert github_config  # guarded by if here
+
+            self._gist_token = github_config["token"]
+            self._gist_timeout = github_config["timeout"]
+            # tokens bucket for gist posting: {paste_id: token\ntoken}
+            self.__tokens_bucket: dict[str, str] = {}
+            self.__token_lock = asyncio.Lock()
+            self.__token_task = asyncio.create_task(self._token_task())
+
+    async def _token_task(self) -> None:
+        # won't run unless pre-reqs are met in __init__
+        while True:
+            if self.__tokens_bucket:
+                async with self.__token_lock:
+                    await self._post_gist_of_tokens()
+
+            await asyncio.sleep(self._gist_timeout)
+
+    def _handle_discord_tokens(self, *bodies: dict[str, str], paste_id: str) -> None:
+        formatted_bodies = "\n".join(b["content"] for b in bodies)
+
+        tokens = list(DISCORD_TOKEN_REGEX.finditer(formatted_bodies))
+
+        if not tokens:
+            return
+
+        tokens = "\n".join([m[0] for m in tokens])
+        self.__tokens_bucket[paste_id] = tokens
+
+    async def _post_gist_of_tokens(self) -> None:
+        assert self.app.session  # guarded in caller
+        json_payload: PostGist = {
+            "description": "MystBin found these Discord tokens in a public paste, and posted them here to invalidate them. If you intended to share these, please apply a password to the paste.",
+            "files": {},
+            "public": True,
+        }
+
+        github_headers = {
+            "Accept": "application/vnd.github+json",
+            "Authorization": f"Bearer {self._gist_token}",
+            "X-GitHub-Api-Version": "2022-11-28",
+        }
+
+        for paste_id, tokens in self.__tokens_bucket.items():
+            filename = str(datetime.datetime.now(datetime.UTC)) + f"/{paste_id}-tokens.txt"
+            json_payload["files"][filename] = {"content": tokens}
+
+        await self.app.session.post("https://api.github.com/gists", headers=github_headers, json=json_payload)
+
+        self.__tokens_bucket = {}
```

Comment:
Should swap these so tokens that are added while the request is in flight are not discarded.
```suggestion
        self.__tokens_bucket = {}

        await self.app.session.post("https://api.github.com/gists", headers=github_headers, json=json_payload)
```